✨feat(awsmachinepool): custom lifecyclehooks for machinepools

sebltm opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind regression
/kind support
-->

**What this PR does / why we need it**:

This PR adds to the v1beta2 definition for the ``AWSMachinePool`` and ``AWSManagedMachinePool`` with a new field ``lifecycleHooks`` which is a list of:
```yaml
name: <the name of the lifecycle hook>
notificationTargetARN: <ARN of resource where to send the lifecycle event; optional>
roleARN: <ARN of role to be used when sending notifications; optional>
lifecycleTransition: <autoscaling:EC2_INSTANCE_LAUNCHING/EC2_INSTANCE_TERMINATING>
heartbeatTimeout: <duration of the heartbeat timeout; optional>
defaultResult: <CONTINUE/ABANDON; optional>
notificationMetadata: <some metadata to add to the notification; optional>
```

The matching webhooks are updated to validate the lifecycle hooks as they are added to the Custom Resource.
The matching reconcilers are updated to enable reconciling those lifecycle hooks: if the lifecycle hook is present in the Custom Resource but not in the cloud, it is created. And if there is a lifecycle hook present in the cloud but not declared in the Custom Resource then it is removed.

<!-- Enter a description of the change and why this change is needed -->

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4020

AWS supports Lifecycle Hooks before/after performing certain actions on an ASG. For example, before scaling in (removing) a node, the ASG can publish an event in an SQS queue which can them be consumed by the node-termination-handler to ensure its proper removal from Kubernetes (it will cordon, drain the node and wait for a period of time for applications to be removed before allowing the Autoscaling Group to terminate the instance).

This allows Kubernetes or other components to be aware of the node's lifecycle and take appropriate actions

**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [x] squashed commits
- [ ] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests
- [x] adds or updates e2e tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Adding support for custom Lifecycle Hooks in AWSMachinePools for external hooks (e.g support for the aws-node-termination-handler with SQS)
```

AndiDog commented:
I have two requests before getting to the review:

- Neither title nor PR description describe the change. Lifecycle hooks and reacting to node shutdown is great – but what is _this PR doing and achieving_? Also, the release note entry in the PR template must be filled.
- You're moving lots of code. Please revert those changes as much as possible so the PR becomes reviewable. Refactoring and file renames can be done separately.

AndiDog commented:
/assign

sebltm commented:
@AndiDog sorry I hadn't cleaned up the PR, I didn't know if it would get some traction :)
I've updated the PR, updated the description. Let me know if it looks good, I'll write some docs and add release notes

sebltm commented:
@AndiDog let me know if this looks good or if there's anything else I should take a look at :)

AndiDog commented:
/ok-to-test

sebltm commented:
@AndiDog could you help me find someone to review this PR? I've posted a couple of times in Slack
The point of the PR isn't just to cater to node termination scenarios, but to enable the flexibility to run actions tied to the node lifecycle (e.g run a lambda to create certain resources or clean up resources on node scale-in/scale-out or when nodes enter or leave the warm pool)

fiunchinho commented on config/crd/bases/infrastructure.cluster.x-k8s.io_awsmanagedmachinepools.yaml:
```diff
@@ -846,6 +846,54 @@ spec:
+                        The ARN of the IAM role that allows the Auto Scaling group to publish to the
+                        specified notification target.
+                      type: string
+                  required:
+                  - lifecycleTransition
```
From the validation webhook, it looks like the `name` field is also required. May be better to add it here as well.

fiunchinho commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -189,13 +195,23 @@ func (r *AWSManagedMachinePoolReconciler) Reconcile(ctx context.Context, req ctr
 	}
 
-	return ctrl.Result{}, r.reconcileNormal(ctx, machinePoolScope, managedControlPlaneScope)
+	infraCluster, err := r.getInfraCluster(ctx, log, cluster, machinePoolScope.ManagedMachinePool)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting infra provider cluster or control plane object: %w", err)
+	}
+	if infraCluster == nil {
+		log.Info("AWSCluster or AWSManagedControlPlane is not ready yet")
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, r.reconcileNormal(ctx, machinePoolScope, managedControlPlaneScope, infraCluster)
```
I was wondering if we need to get the infra cluster again. Isn't the `infraCluster` already a property of `machinePoolScope`? Look at line 174.

AndiDog commented on exp/api/v1beta2/awsmachinepool_webhook.go:
```diff
@@ -116,6 +119,37 @@ func (r *AWSMachinePool) validateSpotInstances() field.ErrorList {
+			allErrs = append(allErrs, field.Required(field.NewPath("spec.lifecycleHooks.notificationTargetARN"), "NotificationTargetARN is required if RoleARN is provided"))
+		}
+		if hook.LifecycleTransition != LifecycleTransitionInstanceLaunch && hook.LifecycleTransition != LifecycleTransitionInstanceTerminate {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec.lifecycleHooks.lifecycleTransition"), hook.LifecycleTransition, "LifecycleTransition must be either EC2_INSTANCE_LAUNCHING or EC2_INSTANCE_TERMINATING"))
```
```suggestion
			allErrs = append(allErrs, field.Invalid(field.NewPath("spec.lifecycleHooks.lifecycleTransition"), hook.LifecycleTransition, fmt.Sprintf("LifecycleTransition must be either %q or %q", LifecycleTransitionInstanceLaunch, LifecycleTransitionInstanceTerminate)))
```

[same below]

AndiDog commented on exp/api/v1beta2/awsmachinepool_types.go:
```diff
@@ -101,6 +101,10 @@ type AWSMachinePoolSpec struct {
 	// If a process is removed from this list it will automatically be resumed.
 	SuspendProcesses *SuspendProcessesTypes `json:"suspendProcesses,omitempty"`
+
+	// AWSLifecycleHooks specifies lifecycle hooks for the managed node group.
```
```suggestion
	// AWSLifecycleHooks specifies lifecycle hooks for the autoscaling group.
```

AndiDog commented on exp/controllers/awsmachinepool_controller.go:
```diff
@@ -298,6 +298,21 @@ func (r *AWSMachinePoolReconciler) reconcileNormal(ctx context.Context, machineP
 		return nil
 	}
 
+	lifecycleHookScope, err := scope.NewLifecycleHookScope(scope.LifecycleHookScopeParams{
```
`machinePoolScope` should instead provide an interface function like `func (*FooScope) LifecycleHooks() []AWSLifecycleHook` – instead of introducing a new type that covers both EC2 and EKS based clusters in the same "class"

AndiDog commented on exp/controllers/awsmachinepool_controller.go:
```diff
@@ -298,6 +298,21 @@ func (r *AWSMachinePoolReconciler) reconcileNormal(ctx context.Context, machineP
+	}
+
+	if err := reconSvc.ReconcileLifecycleHooks(*lifecycleHookScope, asgsvc); err != nil {
+		r.Recorder.Eventf(machinePoolScope.AWSMachinePool, corev1.EventTypeWarning, "FaileLifecycleHooksReconcile", "Failed to reconcile lifecycle hooks: %v", err)
```
That's an error, not a warning

sebltm replied:
There's no "EventTypeWarning", this is how the other reconciles push events to the MachinePool (see line 284 for the ReconcileLaunchTemplate).
We push an event to the MachinePool so that the users can see the events by looking at the object, and exit the reconciliation loop by throwing an error

AndiDog commented on exp/controllers/awsmachinepool_controller_test.go:
```diff
@@ -163,13 +163,14 @@ func TestAWSMachinePoolReconciler(t *testing.T) {
 		recorder = record.NewFakeRecorder(2)
```
Right now, there's no test covering the new functionality.

We need a non-mocked test, see

```
reconciler.reconcileServiceFactory = nil // use real implementation, but keep EC2 calls mocked (`ec2ServiceFactory`)
```

below where the actual EC2 calls are tested. The test should cover different situations, such as no hooks exist, all hooks exist, some hooks need an update, there's a hook too much which should be removed, ...

AndiDog commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -18,6 +18,7 @@ package controllers
 
 import (
 	"context"
+	"fmt"
```
Which overall cases did you test? Did you include EKS node group testing and are you using it? Asking because I'd like to know which of the cluster types was covered in tests (EKS or non-EKS or both).

AndiDog commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -207,6 +223,7 @@ func (r *AWSManagedMachinePoolReconciler) reconcileNormal(
 
 	ekssvc := eks.NewNodegroupService(machinePoolScope)
 	ec2svc := r.getEC2Service(ec2Scope)
+	asgsvc := r.getASGService(clusterScope)
```
Only a feeling, didn't check in detail: shouldn't `ekssvc` have what's needed? We didn't have an `asgsvc` for EKS code before, so why should it be needed now?

Probably related to interfaces again – probably `GetLifecycleHooks(scope scope.LifecycleHookScope)` rather fits into a new interface implemented _by_ the (managed) machine pool scope type?

sebltm replied:
So I tried to keep code duplication to a minimum by having the Lifecycle Hook reconciliation in one place since it should be identical between a ManagedMachinePool and a MachinePool (since both are using ASG in the backend), and therefore this lived in the asgsvc since it wouldn't make any sense to have the ekssvc into the MachinePool controller.
I'll change it to have the same logic in both the EKS NodeGroup Service (which reconciles the ManagedMachinePool) and the ASG Service (which is used to reconcile the MachinePool)

AndiDog replied:
It's now all in `pkg/cloud/services/eks/nodegroup.go` and that looks great, thank you!

AndiDog commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -349,6 +381,73 @@ func (r *AWSManagedMachinePoolReconciler) getEC2Service(scope scope.EC2Scope) se
 	return ec2.NewService(scope)
 }
+
+func (r *AWSManagedMachinePoolReconciler) getInfraCluster(ctx context.Context, log *logger.Logger, cluster *clusterv1.Cluster, awsMachinePool *expinfrav1.AWSManagedMachinePool) (scope.EC2Scope, error) {
```
Already mentioned in another comment – this should already be part of the scope and not re-fetched. Neither should EKS code log anything about `AWSCluster` since that's for non-EKS.

AndiDog commented on pkg/cloud/services/autoscaling/lifecyclehook.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// GetLifecycleHooks returns the lifecycle hooks for the given AutoScalingGroup after retrieving them from the AWS API.
+func (s *Service) GetLifecycleHooks(scope scope.LifecycleHookScope) ([]*expinfrav1.AWSLifecycleHook, error) {
```
This could rather be called `DescribeLifecycleHooks` since it's doing an AWS call and otherwise could be confused with `GetLifecycleHooks` that "only" returns `{AWSMachinePool,...}.spec.lifecycleHooks`.

AndiDog commented on pkg/cloud/services/ec2/launchtemplate.go:
```diff
@@ -276,8 +276,8 @@ func updateMachinePoolAnnotation(lts scope.LaunchTemplateScope, annotation, cont
-// tagsChanged determines which tags to delete and which to add.
-func tagsChanged(annotation map[string]interface{}, src map[string]string) (bool, map[string]string, map[string]string, map[string]interface{}) {
+// TagsChanged determines which tags to delete and which to add.
+func TagsChanged(annotation map[string]interface{}, src map[string]string) (bool, map[string]string, map[string]string, map[string]interface{}) {
```
Unrelated change? Seems not needed to make this public – the function is only used in this file.

AndiDog commented on pkg/cloud/services/ec2/lifecyclehook.go:
```diff
@@ -0,0 +1,91 @@
+	}
+	for _, hook := range hooks {
+		found := false
+		for _, definedHook := range scope.GetLifecycleHooks() {
```
```suggestion
		for _, definedHook := range lifecyleHooks {
```

AndiDog commented on pkg/cloud/services/ec2/lifecyclehook.go:
```diff
@@ -0,0 +1,91 @@
+			}
+		}
+		if !found {
+			scope.Info("Deleting lifecycle hook", "hook", hook.Name)
```
```suggestion
			scope.Info("Deleting extraneous lifecycle hook", "hook", hook.Name)
```

AndiDog commented on pkg/cloud/services/ec2/lifecyclehook.go:
```diff
@@ -0,0 +1,91 @@
+		}
+	}
+
+	conditions.MarkTrue(scope.GetMachinePool(), expinfrav1.LifecycleHookExistsCondition)
```
Also `LifecycleHookReadyCondition`? It's never marked as true (or false).

AndiDog commented on pkg/cloud/services/eks/nodegroup.go:
```diff
@@ -150,7 +150,7 @@ func (s *NodegroupService) remoteAccess() (*eks.RemoteAccessConfig, error) {
 	// but just in case we use an empty list to take advantage of the documented
 	// API behavior
-	var sSGs = []string{}
+	sSGs := []string{}
```
Until here, there were quite a few minor, unneeded changes, plus some small improvements, both out of scope for the PR. If you can put the relevant ones into a separate PR and ping me, I'll get them in. Let's please avoid making this already-large PR review slower by this extra content.

sebltm replied:
Sorry these were auto-linted, I missed removing them from the PR, I'll clean those up

AndiDog commented on pkg/cloud/services/interfaces.go:
```diff
@@ -50,6 +50,12 @@ type ASGInterface interface {
 	ResumeProcesses(name string, processes []string) error
 	SubnetIDs(scope *scope.MachinePoolScope) ([]string, error)
+	GetLifecycleHooks(scope scope.LifecycleHookScope) ([]*expinfrav1.AWSLifecycleHook, error)
+	GetLifecycleHook(scope scope.LifecycleHookScope, hook *expinfrav1.AWSLifecycleHook) (*expinfrav1.AWSLifecycleHook, error)
```
```suggestion
	GetLifecycleHook(scope scope.LifecycleHookScope, hookName string) (*expinfrav1.AWSLifecycleHook, error)
```

(minor)

AndiDog reviewed (commented):
Within my company, we managed today to talk about lifecycle hooks and how they could help with several CAPA features, including aws-node-termination-handler which you mentioned. So I'm feeling up to review it in some detail.

sebltm commented:
@AndiDog sorry for merging into the PR, but let me know if this approach looks better to you and I’ll clean this up with a rebase

AndiDog commented on pkg/cloud/services/eks/nodegroup.go:
```diff
@@ -38,8 +38,14 @@ import (
+	"sigs.k8s.io/cluster-api/util/conditions"
 )
 
+var IgnoreLifecycleHooks = map[string]bool{
```
Please add a short comment where these come from (I guess precreated by EKS, but not documented?)

AndiDog commented on pkg/cloud/services/autoscaling/lifecyclehook.go:
```diff
@@ -0,0 +1,185 @@
+	return hooks, nil
+}
+
+// GetLifecycleHook returns a specific lifecycle hook for the given AutoScalingGroup after retrieving it from the AWS API.
+func (s *Service) DescribeLifecycleHook(asgName string, hook *expinfrav1.AWSLifecycleHook) (*expinfrav1.AWSLifecycleHook, error) {
```
```suggestion
// DescribeLifecycleHook returns a specific lifecycle hook for the given AutoScalingGroup after retrieving it from the AWS API.
func (s *Service) DescribeLifecycleHook(asgName string, hookName string) (*expinfrav1.AWSLifecycleHook, error) {
```

AndiDog reviewed (commented):
Second large round. I also largely commented on code structure (which functions go where, common functionality). Not as bad as the number of comments looks like, I hope.

AndiDog commented:
@sebltm Would you like to go ahead with the PR? Let me know if help is needed. My company also wants this for use with node-termination-handler, so I'm happy to help or take over, whichever you prefer.

AndiDog commented:
I took this over because my company is working on using this feature right now, and the PR was almost ready to go.

I've rebased onto `main` and addressed my code style/refactoring review comments. Let's see if tests pass like this, and then someone else can review.

AndiDog commented:
/test pull-cluster-api-provider-aws-e2e

I successfully tested the feature. Only IAM permissions were missing in the right place (doesn't work in the block with `Resource: ["arn:*:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*"]`).

Once review is done and tests pass, I can squash commits.

sebltm commented:
Sorry I've been away, thank you @AndiDog for picking this one up

fiunchinho commented on exp/api/v1beta2/awsmachinepool_webhook_test.go:
```diff
@@ -174,6 +175,33 @@ func TestAWSMachinePoolValidateCreate(t *testing.T) {
 			},
 			wantErr: true,
 		},
+		{
+			name: "Should fail if either roleARN or notifcationARN is set but not both",
+			pool: &AWSMachinePool{
+				Spec: AWSMachinePoolSpec{
+					AWSLifecycleHooks: []AWSLifecycleHook{
+						{
+							RoleARN: aws.String("role-arn"),
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
```
I'd add a case for only setting `roleARN`, and another one only setting `notificationARN`

AndiDog replied:
done

fiunchinho commented on exp/controllers/awsmachinepool_controller.go:
```diff
@@ -298,6 +298,11 @@ func (r *AWSMachinePoolReconciler) reconcileNormal(ctx context.Context, machineP
 		return nil
 	}
 
+	if err := r.reconcileLifecycleHooks(machinePoolScope, asgsvc); err != nil {
```
We have a `ctx` variable in this function, which we don't pass here, but later down the stack we end up creating a `context.TODO()`. May be worth passing the context that we already have. What do you think?

sebltm replied:
@fiunchinho it looks like most of the other interfaces for the ASGInterface and EC2Interface use the same pattern (they get called from places that have context, and they themselves create their own context.TODO context).
We could start breaking the pattern here, it'd be a bit of a divergence to the rest of the code

fiunchinho replied:
Then for consistency sake, it'd be better to follow the same pattern for now. It could be addressed in a different PR later on

AndiDog replied:
I think creating a `context.TODO` might have been a mistake when usage of `*WithContext` AWS SDK functions was introduced. Context should always be specified where possible in order to support timeouts, for instance. Some other interface functions are correctly taking such an argument already.

AndiDog commented:
I noticed that `CreateASG` didn't handle the hooks. Likely, it's best if both are created atomically, so I added this as another commit.

AndiDog commented:
@sebltm I'll try to continue here to bring it through review

AndiDog commented on exp/api/v1beta2/awsmachinepool_webhook.go:
```diff
@@ -133,7 +133,6 @@ func (r *AWSMachinePool) validateAdditionalSecurityGroups() field.ErrorList {
 	}
 	return allErrs
 }
-
```
Done, that empty line was a mistake from merging

AndiDog commented:
/test pull-cluster-api-provider-aws-e2e

Giving it a try, but E2E might be problematic right now.

fiunchinho reviewed (approved):
/lgtm

AndiDog commented:
/test pull-cluster-api-provider-aws-e2e

AndiDog commented:
Test failures will be fixed via https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5242

AndiDog commented:
Rebased onto latest `main` to get test fixes. Let's try again.

/test pull-cluster-api-provider-aws-e2e

AndiDog commented:
Basic tests look green. I'm squashing now so we can run E2E tests and get ready to merge.

AndiDog commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

AndiDog commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

Another rebase was needed. That's what happens when CI becomes green and lots of PRs get merged at once 😆.

fiunchinho commented:
/test pull-cluster-api-provider-aws-e2e-eks

fiunchinho reviewed (approved):
/lgtm

fiunchinho commented:
/lgtm

sebltm commented:
@fiunchinho I’ve had to rebase to fix some merge conflicts from main, could I get a review again?
@luthermonson also looks like we need your review to complete this

fiunchinho reviewed (approved):
/lgtm

fiunchinho commented:
As @AndiDog is in parental leave, could you take a look @nrb @richardcase @Ankitasw @dlipovetsky ?

fiunchinho commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

AndiDog commented:
/test pull-cluster-api-provider-aws-e2e-eks
/lgtm

damdo commented:
/assign @nrb @richardcase @Ankitasw @dlipovetsky

richardcase commented:
/test pull-cluster-api-provider-aws-e2e

fiunchinho commented:
/retest

nrb commented:
/approve

This will probably introduce merge conflicts on #5447 due to the validation webhook changes.
